### PR TITLE
Core: Implement BaseMetastoreCatalog.registerTable()

### LINKED ISSUE #4995: Nessie: Implement Catalog.registerTable() for NessieCatalog
To migrate tables to Nessie catalog from other catalogs, we can make use of the `Catalog.registerTable()` interface. 

Implementation will be similar to `HiveCatalog.registerTable()`.
By default, tables will be registered to the default branch if the table identifier doesn't specify any branch information.  
If the branch information is provided, register the table to that particular Nessie branch (assumes that branch should exist).

cc: @snazy 

@Mehul2500 will be working on this.

### PR DESCRIPTION
**Pull request by Mehul2500**
For migrating tables between different catalogs, need is there for a registerTable() function on destination catalog side.
Thus, planned to implement the registerTable() function in BaseMetastoreCatalog.

Closes #4995

### CODE REVIEW

File: nessie/src/test/java/org/apache/iceberg/nessie/TestNessieTable.java [hunk 1]
```diff
@@ -356,6 +360,107 @@ public void testListTables() {
...
+  @Test
+  public void testRegisterTableWithGivenBranch() {
+    List<String> metadataVersionFiles = metadataVersionFiles(TABLE_NAME);
+    Assertions.assertThat(1).isEqualTo(metadataVersionFiles.size());
+    catalog.dropTable(TableIdentifier.of(DB_NAME, "`" + TABLE_NAME + "@main`"), false);
+    catalog.registerTable(
+            TableIdentifier.of(DB_NAME, "`" + TABLE_NAME + "@main`"),
+            "file:" + metadataVersionFiles.get(0));
+    Table newTable = catalog.loadTable(TableIdentifier.of(DB_NAME, "`" + TABLE_NAME + "@main`"));
+    Assertions.assertThat(newTable).isNotNull();
```
Comment (ajantha-bhat) [inline nessie/src/test/java/org/apache/iceberg/nessie/TestNessieTable.java:372]: can you extract line 372 to 376 in a common function and reuse in other the testcases?

File: nessie/src/test/java/org/apache/iceberg/nessie/TestNessieTable.java [hunk 2]
```diff
@@ -356,6 +360,107 @@ public void testListTables() {
...
+    Assertions.assertThat(catalog.tableExists(TableIdentifier.of(DB_NAME, "`" + TABLE_NAME + "@main`"))).isTrue();
+  }
+
+  @Test
+  public void testRegisterTableBranchNotPresent() {
+    List<String> metadataVersionFiles = metadataVersionFiles(TABLE_NAME);
+    Assertions.assertThat(1).isEqualTo(metadataVersionFiles.size());
+
+    Assertions.assertThatThrownBy(
+                    () -> catalog.registerTable(
```
Comment (ajantha-bhat) [inline nessie/src/test/java/org/apache/iceberg/nessie/TestNessieTable.java:385]: please recheck the indentation in all newly added code.

File: nessie/src/test/java/org/apache/iceberg/nessie/TestNessieTable.java [hunk 3]
```diff
@@ -356,6 +360,107 @@ public void testListTables() {
     Assertions.assertThat(catalog.tableExists(TABLE_IDENTIFIER)).isTrue();
   }
 
+  @Test
+  public void testRegisterTableWithGivenBranch() {
+    List<String> metadataVersionFiles = metadataVersionFiles(TABLE_NAME);
+    Assertions.assertThat(1).isEqualTo(metadataVersionFiles.size());
+    catalog.dropTable(TableIdentifier.of(DB_NAME, "`" + TABLE_NAME + "@main`"), false);
```
Comment (ajantha-bhat) [inline nessie/src/test/java/org/apache/iceberg/nessie/TestNessieTable.java:367]: please validate the return values drop table, tablexist, registerTable by assert

File: nessie/src/test/java/org/apache/iceberg/nessie/TestNessieTable.java [hunk 4]
```diff
@@ -356,6 +360,107 @@ public void testListTables() {
...
+    Assertions.assertThat(1).isEqualTo(metadataVersionFiles.size());
+
+    Assertions.assertThatThrownBy(() -> catalog.registerTable(TABLE_IDENTIFIER, "file:" + metadataVersionFiles.get(0)))
+            .isInstanceOf(AlreadyExistsException.class)
+            .hasMessage(
+                    "Table already exists: db.tbl");
+  }
+
+  @Test
+  public void testRegisterTableTagName() throws NessieConflictException, NessieNotFoundException {
```
Comment (ajantha-bhat) [inline nessie/src/test/java/org/apache/iceberg/nessie/TestNessieTable.java:419]: you can combine all 3 negative scenario test case (that is checking Assertions.assertThatThrownBy) into one.

File: nessie/src/test/java/org/apache/iceberg/nessie/TestNessieTable.java [hunk 5]
```diff
@@ -356,6 +360,107 @@ public void testListTables() {
...
+            "file:" + metadataVersionFiles1.get(0));
+    Table newTable1 = catalog.loadTable(TableIdentifier.of(DB_NAME, "`" + TABLE_NAME + "@branch1`"));
+    Assertions.assertThat(newTable1).isNotNull();
+    TableOperations ops1 = ((HasTableOperations) newTable1).operations();
+    String metadataLocation1 = ((NessieTableOperations) ops1).currentMetadataLocation();
+    Assertions.assertThat("file:" + metadataVersionFiles1.get(0)).isEqualTo(metadataLocation1);
+    Assertions.assertThat(catalog.tableExists(TableIdentifier.of(DB_NAME, "`" + TABLE_NAME + "@branch1`"))).isTrue();
+  }
+
   private String getTableBasePath(String tableName) {
```
Comment (ajantha-bhat) [inline nessie/src/test/java/org/apache/iceberg/nessie/TestNessieTable.java:464]: Add a test case with invalid identifiers, non-null metadata path but metadata file not exist in that path, null metadata location.

File: core/src/main/java/org/apache/iceberg/BaseMetastoreCatalog.java [hunk 6]
```diff
@@ -65,6 +66,22 @@ public Table loadTable(TableIdentifier identifier) {
...
+
+    // Throw an exception if this table already exists in the catalog.
+    if (tableExists(identifier)) {
+      throw new org.apache.iceberg.exceptions.AlreadyExistsException("Table already exists: %s", identifier);
+    }
+
+    TableOperations ops = newTableOps(identifier);
+    InputFile metadataFile = ops.io().newInputFile(metadataFileLocation);
+    TableMetadata metadata = TableMetadataParser.read(ops.io(), metadataFile);
+    ops.commit(null, metadata);
```
[the remark this answers is not in the document]
Comment (Mehul2500) [reply core/src/main/java/org/apache/iceberg/BaseMetastoreCatalog.java:81]: It shows up, 'Cannot call commit on temporary table operations', thus continuing with the previous code statement,i.e., line 81.

File: core/src/main/java/org/apache/iceberg/BaseMetastoreCatalog.java [hunk 7]
```diff
@@ -65,6 +66,24 @@ public Table loadTable(TableIdentifier identifier) {
...
+
+    return new BaseTable(ops, identifier.toString());
+  }
```
Comment (ajantha-bhat) [inline core/src/main/java/org/apache/iceberg/BaseMetastoreCatalog.java:86]: please have a new line here.

File: core/src/main/java/org/apache/iceberg/BaseMetastoreCatalog.java [hunk 8]
```diff
@@ -65,6 +66,24 @@ public Table loadTable(TableIdentifier identifier) {
     return result;
   }
 
+  @Override
+  public org.apache.iceberg.Table registerTable(TableIdentifier identifier, String metadataFileLocation) {
+    Preconditions.checkArgument(isValidIdentifier(identifier), "Invalid identifier: %s", identifier);
```
Comment (ajantha-bhat) [inline core/src/main/java/org/apache/iceberg/BaseMetastoreCatalog.java:71]: Need to have a null check for the identifier?

File: core/src/main/java/org/apache/iceberg/BaseMetastoreCatalog.java (hunk 6 above)
[the remark this answers is not in the document]
Comment (ajantha-bhat) [reply core/src/main/java/org/apache/iceberg/BaseMetastoreCatalog.java:83]: ok. Intersting.

File: nessie/src/test/java/org/apache/iceberg/nessie/TestNessieTable.java [hunk 9]
```diff
@@ -356,6 +361,107 @@ public void testListTables() {
     Assertions.assertThat(catalog.tableExists(TABLE_IDENTIFIER)).isTrue();
   }
 
+  private void testRegister(TableIdentifier identifier, List<String> metadataVersionFiles) {
```
Comment (ajantha-bhat) [inline nessie/src/test/java/org/apache/iceberg/nessie/TestNessieTable.java:364]: better to accept single metadata file instead of list of files

File: nessie/src/test/java/org/apache/iceberg/nessie/TestNessieTable.java [hunk 10]
```diff
@@ -356,6 +361,107 @@ public void testListTables() {
...
+    Assertions.assertThat(catalog.dropTable(identifier, false)).isFalse();
+    Assertions.assertThat(catalog.registerTable(
+            identifier,
```
Comment (ajantha-bhat) [inline nessie/src/test/java/org/apache/iceberg/nessie/TestNessieTable.java:367]: Indentations are still not correct. Please recheck the newly added codes again.

File: nessie/src/test/java/org/apache/iceberg/nessie/TestNessieTable.java [hunk 11]
```diff
@@ -356,6 +361,107 @@ public void testListTables() {
...
+            "file:" + metadataVersionFiles.get(0))).isNotNull();
+    Table newTable = catalog.loadTable(identifier);
+    Assertions.assertThat(newTable).isNotNull();
+    TableOperations ops = ((HasTableOperations) newTable).operations();
+    String metadataLocation = ((NessieTableOperations) ops).currentMetadataLocation();
+    Assertions.assertThat("file:" + metadataVersionFiles.get(0)).isEqualTo(metadataLocation);
+    Assertions.assertThat(catalog.tableExists(identifier)).isTrue();
```
Comment (ajantha-bhat) [inline nessie/src/test/java/org/apache/iceberg/nessie/TestNessieTable.java:374]: load table is called above. So, this check is not needed as both are same functionality.

File: nessie/src/test/java/org/apache/iceberg/nessie/TestNessieTable.java [hunk 12]
```diff
@@ -356,6 +361,107 @@ public void testListTables() {
...
+            () -> catalog.registerTable(TABLE_IDENTIFIER, "file:" + metadataVersionFiles.get(0) + "invalidName"))
+        .isInstanceOf(NotFoundException.class);
+  }
+
+  @Test
+  public void testRegisterTableWithDefaultBranch() {
+    List<String> metadataVersionFiles = metadataVersionFiles(TABLE_NAME);
+    Assertions.assertThat(1).isEqualTo(metadataVersionFiles.size());
+    Assertions.assertThat(catalog.dropTable(TABLE_IDENTIFIER)).isTrue();
+    Assertions.assertThat(catalog.registerTable(TABLE_IDENTIFIER, "file:" + metadataVersionFiles.get(0))).isNotNull();
```
[the remark this answers is not in the document]
Comment (Mehul2500) [reply nessie/src/test/java/org/apache/iceberg/nessie/TestNessieTable.java:423]: We cannot do so as the line wherein we assert for dropTable in this particular test case is different to the assertion we implemented in testRegister(), because of the dropTable backend bugs.

File: nessie/src/test/java/org/apache/iceberg/nessie/TestNessieTable.java [hunk 13]
```diff
@@ -356,6 +361,107 @@ public void testListTables() {
...
+  }
+
+  @Test
+  public void testRegisterTableWithGivenBranch() {
+    List<String> metadataVersionFiles = metadataVersionFiles(TABLE_NAME);
+    Assertions.assertThat(1).isEqualTo(metadataVersionFiles.size());
+    ImmutableTableReference tableReference =
+            ImmutableTableReference.builder().reference("main").name(TABLE_NAME).build();
+    TableIdentifier identifier = TableIdentifier.of(DB_NAME, tableReference.toString());
+    testRegister(identifier, metadataVersionFiles);
```
[the remark this answers is not in the document]
Comment (ajantha-bhat) [reply nessie/src/test/java/org/apache/iceberg/nessie/TestNessieTable.java:384]: ok

File: nessie/src/test/java/org/apache/iceberg/nessie/TestNessieTable.java (hunk 12 above)
[the remark this answers is not in the document]
Comment (ajantha-bhat) [reply nessie/src/test/java/org/apache/iceberg/nessie/TestNessieTable.java:423]: please rebase on #5033

File: core/src/main/java/org/apache/iceberg/BaseMetastoreCatalog.java [hunk 14]
```diff
@@ -65,6 +66,26 @@ public Table loadTable(TableIdentifier identifier) {
     return result;
   }
 
+  @Override
+  public org.apache.iceberg.Table registerTable(TableIdentifier identifier, String metadataFileLocation) {
+    Preconditions.checkArgument(
+        isValidIdentifier(identifier) && identifier != null, "Invalid identifier: %s", identifier);
```
Comment (pvary) [inline core/src/main/java/org/apache/iceberg/BaseMetastoreCatalog.java:72]: it would be more natural to do the null check first

File: nessie/src/main/java/org/apache/iceberg/nessie/NessieCatalog.java [hunk 15]
```diff
@@ -196,7 +194,16 @@ public List<TableIdentifier> listTables(Namespace namespace) {
 
   @Override
   public boolean dropTable(TableIdentifier identifier, boolean purge) {
-    return client.dropTable(identifier, purge);
+    TableReference tableReference = parseTableReference(identifier);
```
[the remark this answers is not in the document]
Comment (pvary) [reply nessie/src/main/java/org/apache/iceberg/nessie/NessieCatalog.java:197]: Probably we should not mix PRs.
Let's commit the changes one-by-one

File: nessie/src/test/java/org/apache/iceberg/nessie/TestNessieTable.java [hunk 16]
```diff
@@ -356,6 +360,81 @@ public void testListTables() {
     Assertions.assertThat(catalog.tableExists(TABLE_IDENTIFIER)).isTrue();
   }
 
+  private void testRegister(TableIdentifier identifier, String metadataVersionFiles) {
```
Comment (pvary) [inline nessie/src/test/java/org/apache/iceberg/nessie/TestNessieTable.java:363]: Could we separate these tests for Nessie specific and general ones?
Could we run the "general" tests on every catalog, and move the `testRegisterTable`, and `testRegisterExistingTable` from `HiveTableTest` to this general place?

File: nessie/src/main/java/org/apache/iceberg/nessie/NessieCatalog.java (hunk 15 above)
[the remark this answers is not in the document]
Comment (ajantha-bhat) [reply nessie/src/main/java/org/apache/iceberg/nessie/NessieCatalog.java:197]: @pvary: please help merging #5033, So that we can rebase this.

File: nessie/src/test/java/org/apache/iceberg/nessie/TestNessieTable.java (hunk 16 above)
Comment (Mehul2500) [reply nessie/src/test/java/org/apache/iceberg/nessie/TestNessieTable.java:363]: These test cases are specific to Nessie.

Comment (pvary) [reply nessie/src/test/java/org/apache/iceberg/nessie/TestNessieTable.java:363]: If I understand correctly, the PR moves the `HiveCatalog.registerTable` to the `BaseMetastoreCatalog` so it will be available for every `Catalog` which is inheriting from `BaseMetastoreCatalog`. So we should have a test for testing all of the catalogs that implement the `registerTable` that the method is working as expected.
I understand that there are some Nessie specific test cases, but I think the basic functionality should be available and tested for all of the catalogs.

Comment (Mehul2500) [reply nessie/src/test/java/org/apache/iceberg/nessie/TestNessieTable.java:391]: I will add the said test-cases and would ping you then, thanks.

Comment (pvary) [reply nessie/src/test/java/org/apache/iceberg/nessie/TestNessieTable.java:391]: I think we still miss:
- DynamoDbCatalog
- GlueCatalog

I do not like the duplicated test code, and I think it would be good to have a general Catalog API test for checking the functionality of the new Catalog implementations. Something like `TestCatalog` but for all/most of the catalog API methods. We can do it in a different PR, but it would be good to have them. The test for `registerTable()` is a good candidate for these common tests.

Comment (Mehul2500) [reply nessie/src/test/java/org/apache/iceberg/nessie/TestNessieTable.java:391]: I have put in test cases for Glue and DynamoDb as "Ignore" because I currently do not have an AWS account for testing them.
Also, I feel we could address the Catalog API test in a different PR.

File: aws/src/integration/java/org/apache/iceberg/aws/dynamodb/TestDynamoDbCatalog.java [hunk 17]
```diff
@@ -295,6 +299,30 @@ public void testDropNamespace() {
     Assert.assertFalse("namespace must not exist", response.hasItem());
   }
 
+  @Ignore
+  public void testRegisterTable() {
+    TableIdentifier identifier = TableIdentifier.of("a", "t1");
```
Comment (ajantha-bhat) [inline aws/src/integration/java/org/apache/iceberg/aws/dynamodb/TestDynamoDbCatalog.java:304]: probably gets a namespace doesn't exist error. Need to create a namespace first? Also good to use `genRandomName()` as in other testcases.

File: aws/src/integration/java/org/apache/iceberg/aws/dynamodb/TestDynamoDbCatalog.java [hunk 18]
```diff
@@ -295,6 +299,30 @@ public void testDropNamespace() {
...
+    TableOperations ops = ((HasTableOperations) registeringTable).operations();
+    String metadataLocation = ((DynamoDbTableOperations) ops).currentMetadataLocation();
+    Assertions.assertThat(catalog.registerTable(identifier, metadataLocation)).isNotNull();
+    Table newTable = catalog.loadTable(identifier);
+    Assertions.assertThat(newTable).isNotNull();
+  }
+
+  @Ignore
+  public void testRegisterExistingTable() {
+    catalog.createTable(TableIdentifier.of("a", "t1"), SCHEMA);
```
Comment (ajantha-bhat) [inline aws/src/integration/java/org/apache/iceberg/aws/dynamodb/TestDynamoDbCatalog.java:317]: make a table identifier once and use in all 3 places.

and same comments as above.

File: aws/src/integration/java/org/apache/iceberg/aws/glue/TestGlueCatalogTable.java [hunk 19]
```diff
@@ -433,4 +435,28 @@ public void testTablePropsDefinedAtCatalogLevel() {
...
+
+  @Ignore
+  public void testRegisterTable() {
+    String namespace = createNamespace();
+    String tableName = getRandomName();
+    createTable(namespace, tableName);
+    Table table = glueCatalog.loadTable(TableIdentifier.of(namespace, tableName));
+    String metadataLocation = ((BaseTable) table).operations().current().metadataFileLocation();
+    glueCatalog.dropTable(TableIdentifier.of(namespace, tableName), false);
+    glueCatalog.registerTable(TableIdentifier.of(namespace, tableName), metadataLocation);
```
Comment (ajantha-bhat) [inline aws/src/integration/java/org/apache/iceberg/aws/glue/TestGlueCatalogTable.java:447]: need to validate the results and test load table?

File: core/src/test/java/org/apache/iceberg/hadoop/TestHadoopCatalog.java [hunk 20]
```diff
@@ -589,4 +593,31 @@ public void testTablePropsDefinedAtCatalogLevel() throws IOException {
...
+  public void testRegisterTable() throws IOException {
+    TableIdentifier identifier = TableIdentifier.of("a", "t1");
+    TableIdentifier identifier2 = TableIdentifier.of("a", "t2");
+    HadoopCatalog catalog = hadoopCatalog();
+    catalog.createTable(identifier, SCHEMA);
+    Table registeringTable = catalog.loadTable(identifier);
+    TableOperations ops = ((HasTableOperations) registeringTable).operations();
+    String metadataLocation = ((HadoopTableOperations) ops).current().metadataFileLocation();
+    Assertions.assertThat(catalog.registerTable(identifier2, metadataLocation)).isNotNull();
+    Table newTable = catalog.loadTable(identifier2);
```
Comment (ajantha-bhat) [inline core/src/test/java/org/apache/iceberg/hadoop/TestHadoopCatalog.java:607]: nit: no need for the variable if it is used only once. Can be

`Assertions.assertThat(catalog.loadTable(identifier2)).isNotNull();`

File: aws/src/integration/java/org/apache/iceberg/aws/glue/TestGlueCatalogTable.java [hunk 21]
```diff
@@ -433,4 +435,28 @@ public void testTablePropsDefinedAtCatalogLevel() {
...
+  @Ignore
+  public void testRegisterTableAlreadyExists() {
+    String namespace = createNamespace();
+    String tableName = getRandomName();
+    createTable(namespace, tableName);
+    Table table = glueCatalog.loadTable(TableIdentifier.of(namespace, tableName));
+    String metadataLocation = ((BaseTable) table).operations().current().metadataFileLocation();
+    AssertHelpers.assertThrows("Should fail to register to an existing Glue table",
+        AlreadyExistsException.class,
+        "already exists in Glue",
```
Comment (ajantha-bhat) [inline aws/src/integration/java/org/apache/iceberg/aws/glue/TestGlueCatalogTable.java:459]: you many not get this error message `"already exists in Glue"`

File: nessie/src/test/java/org/apache/iceberg/nessie/TestNessieTable.java (hunk 16 above)
Comment (Mehul2500) [reply nessie/src/test/java/org/apache/iceberg/nessie/TestNessieTable.java:391]: I got my AWS account setup, and have amended the test cases in Glue and DynamoDb Catalog, @pvary please have a look over the test cases.

File: nessie/src/test/java/org/apache/iceberg/nessie/TestNessieTable.java [hunk 22]
```diff
@@ -385,6 +388,83 @@ public void testDropTable() throws IOException {
...
+        .hasMessage("Nessie ref 'default' does not exist");
+    // Case 2: Table Already Exists
+    Assertions.assertThatThrownBy(() -> catalog.registerTable(TABLE_IDENTIFIER, "file:" + metadataVersionFiles.get(0)))
+        .isInstanceOf(AlreadyExistsException.class)
+        .hasMessage("Table already exists: db.tbl");
+    // Case 3: Registering using a tag
+    api.createReference().sourceRefName(BRANCH).reference(Tag.of("tag_1", catalog.currentHash())).create();
+    Assertions.assertThatThrownBy(
+        () -> catalog.registerTable(
+            TableIdentifier.of(DB_NAME, "`" + TABLE_NAME + "`@tag_1"),
```
Comment (ajantha-bhat) [inline nessie/src/test/java/org/apache/iceberg/nessie/TestNessieTable.java:432]: nit: can use `TableReference` here for tag also.

**Review by ajantha-bhat — APPROVED** [review]
LGTM

- Manually verified the integration tests with my AWS account. 
- Agree with @pvary that test cases can be refactored into common class and other class should extend it to avoid duplication. But this can be done in follow up PR.

**Comment by Mehul2500** [comment]
@pvary and @rdblue could you please review this PR.

File: core/src/main/java/org/apache/iceberg/BaseMetastoreCatalog.java [hunk 23]
```diff
@@ -65,6 +66,26 @@ public Table loadTable(TableIdentifier identifier) {
...
   }
 
+  @Override
+  public Table registerTable(TableIdentifier identifier, String metadataFileLocation) {
+    Preconditions.checkArgument(
+        (identifier != null) && isValidIdentifier(identifier), "Invalid identifier: %s", identifier);
+    Preconditions.checkArgument(metadataFileLocation != null && !metadataFileLocation.isEmpty(),
+        "Cannot register an empty metadata file location as a table");
+
+    // Throw an exception if this table already exists in the catalog.
```
Comment (nastra) [inline core/src/main/java/org/apache/iceberg/BaseMetastoreCatalog.java:76]: I think it makes sense to adjust the impl here to what Ryan suggested in https://github.com/apache/iceberg/pull/4099/files#r805404106

File: nessie/src/test/java/org/apache/iceberg/nessie/TestNessieTable.java [hunk 24]
```diff
@@ -385,6 +388,89 @@ public void testDropTable() throws IOException {
...
+    List<String> metadataVersionFiles = metadataVersionFiles(TABLE_NAME);
+    Assertions.assertThat(1).isEqualTo(metadataVersionFiles.size());
+    ImmutableTableReference tableReference =
+        ImmutableTableReference.builder().reference("main").name(TABLE_NAME).build();
+    TableIdentifier identifier = TableIdentifier.of(DB_NAME, tableReference.toString());
+    testRegister(identifier, metadataVersionFiles.get(0));
+  }
+
+  @Test
+  public void testRegisterTableNegativeScenarios() throws NessieConflictException, NessieNotFoundException {
```
Comment (dimas-b) [inline nessie/src/test/java/org/apache/iceberg/nessie/TestNessieTable.java:412]: "Negative" in what sense? Would you mind renaming to `testRegisterTableFailureScenarios`? Also, it would be preferable to have a separate test method for each case, IMHO.

File: nessie/src/test/java/org/apache/iceberg/nessie/TestNessieTable.java [hunk 25]
```diff
@@ -385,6 +388,89 @@ public void testDropTable() throws IOException {
     verifyCommitMetadata();
   }
 
+  private void testRegister(TableIdentifier identifier, String metadataVersionFiles) {
```
Comment (dimas-b) [inline nessie/src/test/java/org/apache/iceberg/nessie/TestNessieTable.java:391]: This is a utility method, but it is names like a test method... Would you mind renaming to something like `validateRegister`?

File: nessie/src/test/java/org/apache/iceberg/nessie/TestNessieTable.java (hunk 16 above)
Comment (kbendick) [reply nessie/src/test/java/org/apache/iceberg/nessie/TestNessieTable.java:391]: > I do not like the duplicated test code, and I think it would be good to have a general Catalog API test for checking the functionality of the new Catalog implementations. Something like TestCatalog but for all/most of the catalog API methods. We can do it in a different PR, but it would be good to have them. The test for registerTable() is a good candidate for these common tests.

We do have `CatalogTests`, which is what `TestRESTCatalog` is built off of. It’s nice in that it has some methods for subclasses to declare if the catalog being tested supports certain features, so the tests can exit early if they don’t support an optional feature.

For example if the catalog in question doesn’t support namespace properties.

Might be good to see if that can be made to fit this need or can otherwise get some ideas from it. I believe Nessie also implements these tests.

File: core/src/main/java/org/apache/iceberg/BaseMetastoreCatalog.java (hunk 23 above)
Comment (Mehul2500) [reply core/src/main/java/org/apache/iceberg/BaseMetastoreCatalog.java:76]: A similar comment was posted out by @ajantha-bhat, please refer https://github.com/apache/iceberg/pull/5037#discussion_r898211187
TLDR it gives an error

File: nessie/src/test/java/org/apache/iceberg/nessie/TestNessieTable.java (hunk 24 above)
Comment (Mehul2500) [reply nessie/src/test/java/org/apache/iceberg/nessie/TestNessieTable.java:412]: I have considered the similar approach of putting the Failure Scenarios in one test case as done for `testRenameWithTableReferenceInvalidCase()` in the same file.

[the remark this answers is not in the document]
Comment (Mehul2500) [reply nessie/src/test/java/org/apache/iceberg/nessie/TestNessieTable.java:412]: I have considered the similar approach of putting the Failure Scenarios in one test case as done for testRenameWithTableReferenceInvalidCase() in the same file.

File: nessie/src/test/java/org/apache/iceberg/nessie/TestNessieTable.java (hunk 16 above)
Comment (Mehul2500) [reply nessie/src/test/java/org/apache/iceberg/nessie/TestNessieTable.java:391]: A good suggestion of using `CatalogTests`, but the duplicated test code referred to under these comments above shall not be resolved using `CatalogTests`.
Reason:
1. The duplicated test code is with two of the test cases, i.e., testRegisterTable() and testRegisterExistingTable() in three of the catalog test files.

- Jdbc Catalog
- Ecs Catalog
- Hadoop Catalog

Out of these only Jdbc catalog is one which can use `CatalogTests`, others do not use `CatalogTests`.
Thus, it won't be helpful to consider `CatalogTests` under this comment.

File: core/src/test/java/org/apache/iceberg/jdbc/TestJdbcCatalog.java [hunk 26]
```diff
@@ -639,4 +642,29 @@ public void testConversions() {
     Assert.assertEquals(ns, JdbcUtil.stringToNamespace(nsString));
   }
 
+  @Test
+  public void testRegisterTable() {
```
[the remark this answers is not in the document]
Comment (Mehul2500) [reply core/src/test/java/org/apache/iceberg/jdbc/TestJdbcCatalog.java:646]: Please refer to https://github.com/apache/iceberg/pull/5037#discussion_r921579263

File: nessie/src/test/java/org/apache/iceberg/nessie/TestNessieTable.java (hunk 25 above)
Comment (Mehul2500) [reply nessie/src/test/java/org/apache/iceberg/nessie/TestNessieTable.java]: updated...thank you

File: nessie/src/main/java/org/apache/iceberg/nessie/NessieTableOperations.java [hunk 27]
```diff
@@ -152,7 +152,8 @@ protected void doCommit(TableMetadata base, TableMetadata metadata) {
       }
     }
 
-    String newMetadataLocation = writeNewMetadata(metadata, currentVersion() + 1);
+    String newMetadataLocation = (base == null) && (metadata.metadataFileLocation() != null) ?
+        metadata.metadataFileLocation() : writeNewMetadata(metadata, currentVersion() + 1);
```
[the remark this answers is not in the document]
Comment (Mehul2500) [reply nessie/src/main/java/org/apache/iceberg/nessie/NessieTableOperations.java:156]: I have considered the changes similar to the `HiveTableOperations` file

File: core/src/main/java/org/apache/iceberg/BaseMetastoreCatalog.java [hunk 28]
```diff
@@ -65,6 +66,26 @@ public Table loadTable(TableIdentifier identifier) {
     return result;
   }
 
+  @Override
+  public Table registerTable(TableIdentifier identifier, String metadataFileLocation) {
+    Preconditions.checkArgument(
+        (identifier != null) && isValidIdentifier(identifier), "Invalid identifier: %s", identifier);
```
[the remark this answers is not in the document]
Comment (Mehul2500) [reply core/src/main/java/org/apache/iceberg/BaseMetastoreCatalog.java]: updated...thank you

**Review by pvary — APPROVED** [review]
Let's see if all of us agree, that is is ok to do the Catalog API test in a different PR.
If no objections arrive, we can merge this PR in a few days

**Comment by pvary** [comment]
Merged the PR.
Thanks for the work @Mehul2500 and @ajantha-bhat for the review.

@Mehul2500 if you have time please go ahead with the CatalogAPI test PR.

Thanks,
Peter
